File: src/chapters/nonlinear-equations/content/code/bisection.m
function [c, k] = bisection(f, a, b, tol, max_iter)
% BISECTION  Root of f on [a,b] (requires f(a)*f(b) < 0).
    if nargin < 4, tol = 1e-12; end
    if nargin < 5, max_iter = 200; end
    fa = f(a);
    for k = 1:max_iter
        c = (a + b)/2; fc = f(c);
        if fc == 0 || (b - a)/2 < tol, return; end
        if fa*fc < 0, b = c; else, a = c; fa = fc; end
    end
end

% --- Demo ---
disp(bisection(@(x) x^2 - 2, 1, 2));        % sqrt(2)

File: src/chapters/nonlinear-equations/content/code/bisection.py
def bisection(f, a, b, tol=1e-12, max_iter=200):
    """Bisection root of f on [a, b] (requires f(a)*f(b) < 0)."""
    fa = f(a)
    for k in range(1, max_iter + 1):
        c = (a + b) / 2
        fc = f(c)
        if fc == 0 or (b - a) / 2 < tol:
            return c, k
        if fa * fc < 0:
            b = c
        else:
            a, fa = c, fc
    return (a + b) / 2, max_iter


if __name__ == "__main__":
    f = lambda x: x * x - 2
    print(bisection(f, 1, 2))           # -> sqrt(2)
